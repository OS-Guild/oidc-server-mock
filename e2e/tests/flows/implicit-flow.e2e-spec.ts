import * as querystring from 'querystring';
import * as dotenv from 'dotenv';
import { decode as decodeJWT } from 'jws';
import { chromium, Page, Browser } from 'playwright-chromium';

import type { User, Client } from '../../types';
import users from '../../config/user-configuration.json';
import clients from '../../config/clients-configuration.json';
import { authorizationEndpoint, grants, introspectEndpoint, userInfoEndpoint } from '../../helpers';

const testCases: User[] = users
  .map(u => ({
    ...u,
    toString: function () {
      return this.SubjectId;
    },
  }))
  .sort((u1, u2) => (u1.SubjectId < u2.SubjectId ? -1 : 1));

describe('Implicit Flow', () => {
  let token: string;

  let browser: Browser;
  let page: Page;
  let client: Client;

  beforeAll(async () => {
    dotenv.config();

    browser = await chromium.launch();
    client = clients.find(c => c.ClientId === 'implicit-flow-client-id');
    expect(client).toBeDefined();
  });

  beforeEach(async () => {
    page = await browser.newPage();
  });

  afterEach(async () => {
    await page.close();
  });

  afterAll(async () => {
    await browser.close();
  });

  describe.each(testCases)('- %s -', (user: User) => {
    test('Authorization Endpoint', async () => {
      const parameters = {
        client_id: client.ClientId,
        scope: 'openid profile email some-custom-identity some-app-scope-1',
        response_type: 'id_token token',
        redirect_uri: client.RedirectUris?.[0].replace('*', 'www'),
        state: 'abc',
        nonce: 'xyz',
      };
      const redirectedUrl = await authorizationEndpoint(page, parameters, user, parameters.redirect_uri);
      const hash = redirectedUrl.hash.slice(1);
      const query = querystring.parse(hash);

      const tokenParameter = query['access_token'];
      expect(typeof tokenParameter).toEqual('string');
      token = tokenParameter as string;
      const decodedAccessToken = decodeJWT(token);
      expect(decodedAccessToken).toMatchSnapshot();
    });

    test('UserInfo Endpoint', async () => {
      await userInfoEndpoint(token);
    });

    test('Introspection Endpoint', async () => {
      await introspectEndpoint(token, 'some-app');
    });

    test('Grants', async () => {
      await grants(page, user);
    });
  });
});
