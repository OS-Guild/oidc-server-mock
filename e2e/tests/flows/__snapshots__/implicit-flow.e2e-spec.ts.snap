// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implicit Flow - simple_user - Authorization Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "simple_user",
  "idp": "local",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ]
}
`;

exports[`Implicit Flow - simple_user - Grants 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, shrink-to-fit=no\\">

    <title>IdentityServer4</title>
    
    <link rel=\\"icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    <link rel=\\"shortcut icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    
    <link rel=\\"stylesheet\\" href=\\"/lib/bootstrap/dist/css/bootstrap.min.css\\">
    <link rel=\\"stylesheet\\" href=\\"/css/site.css\\">
</head>
<body>
    

<div class=\\"nav-page\\">
    <nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\">

        <a href=\\"/\\" class=\\"navbar-brand\\">
            <img src=\\"/icon.png\\" class=\\"icon-banner\\">
            IdentityServer4
        </a>

            <ul class=\\"navbar-nav mr-auto\\">
                <li class=\\"nav-item dropdown\\">
                    <a href=\\"#\\" class=\\"nav-link dropdown-toggle\\" data-toggle=\\"dropdown\\">simple_user <b class=\\"caret\\"></b></a>
                    
                    <div class=\\"dropdown-menu\\">
                        <a class=\\"dropdown-item\\" href=\\"/Account/Logout\\">Logout</a>
                    </div>
              </li>
            </ul>
    
    </nav>
</div>

   
    <div class=\\"container body-container\\">
        
<div class=\\"grants-page\\">
    <div class=\\"lead\\">
        <h1>Client Application Permissions</h1>
        <p>Below is the list of applications you have given permission to and the resources they have access to.</p>
    </div>

            <div class=\\"card\\">
                <div class=\\"card-header\\">
                    <div class=\\"row\\">
                        <div class=\\"col-sm-8 card-title\\">
                            <strong>authorization-code-client-id</strong>
                        </div>

                        <div class=\\"col-sm-2\\">
                            <form action=\\"/Grants/Revoke\\" method=\\"post\\">
                                <input type=\\"hidden\\" name=\\"clientId\\" value=\\"authorization-code-client-id\\">
                                <button class=\\"btn btn-danger\\">Revoke Access</button>
                            <input name=\\"__RequestVerificationToken\\" type=\\"hidden\\" value=\\"CfDJ8Eky9VyqqPNLoUcaLSvjcIr4ohQi4vmy7D16PmFcKrzm8VMk2FNKKNO5UV2UE6WldefgchilGnXaetu7WrSBNGybsRvNdxTBZj7bR78_LwoaOWBEau2lhAcLYZs0-qgQXvuphFCK-3TWxSsH8WqstjdteGDZMlk6cP1dlR-kXlVkbbw2ePMjdCwFQyu4alH51w\\"></form>
                        </div>
                    </div>
                </div>
                
                <ul class=\\"list-group list-group-flush\\">
                    <li class=\\"list-group-item\\">
                        <label>Created:</label> 2020-10-15
                    </li>
                        <li class=\\"list-group-item\\">
                            <label>Expires:</label> 2020-10-15
                        </li>
                        <li class=\\"list-group-item\\">
                            <label>Identity Grants</label>
                            <ul>
                                    <li>Your user identifier</li>
                                    <li>User profile</li>
                                    <li>Your email address</li>
                                    <li>some-custom-identity</li>
                            </ul>
                        </li>
                        <li class=\\"list-group-item\\">
                            <label>API Grants</label>
                            <ul>
                                    <li>some-app-scope-1</li>
                            </ul>
                        </li>
                </ul>
            </div>
</div>
    </div>

    <script src=\\"/lib/jquery/dist/jquery.slim.min.js\\"></script>
    <script src=\\"/lib/bootstrap/dist/js/bootstrap.bundle.min.js\\"></script>

    


</body></html>"
`;

exports[`Implicit Flow - simple_user - Introspection Endpoint 1`] = `
{
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "simple_user",
  "idp": "local",
  "amr": "pwd",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Implicit Flow - simple_user - UserInfo Endpoint 1`] = `
Object {
  "sub": "simple_user",
}
`;

exports[`Implicit Flow - user_with_all_claim_types - Authorization Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_all_claim_types",
  "idp": "local",
  "some-app-user-custom-claim": "Oliver's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Oliver's Scope Custom User Claim",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ]
}
`;

exports[`Implicit Flow - user_with_all_claim_types - Grants 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, shrink-to-fit=no\\">

    <title>IdentityServer4</title>
    
    <link rel=\\"icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    <link rel=\\"shortcut icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    
    <link rel=\\"stylesheet\\" href=\\"/lib/bootstrap/dist/css/bootstrap.min.css\\">
    <link rel=\\"stylesheet\\" href=\\"/css/site.css\\">
</head>
<body>
    

<div class=\\"nav-page\\">
    <nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\">

        <a href=\\"/\\" class=\\"navbar-brand\\">
            <img src=\\"/icon.png\\" class=\\"icon-banner\\">
            IdentityServer4
        </a>

            <ul class=\\"navbar-nav mr-auto\\">
                <li class=\\"nav-item dropdown\\">
                    <a href=\\"#\\" class=\\"nav-link dropdown-toggle\\" data-toggle=\\"dropdown\\">user_with_all_claim_types <b class=\\"caret\\"></b></a>
                    
                    <div class=\\"dropdown-menu\\">
                        <a class=\\"dropdown-item\\" href=\\"/Account/Logout\\">Logout</a>
                    </div>
              </li>
            </ul>
    
    </nav>
</div>

   
    <div class=\\"container body-container\\">
        
<div class=\\"grants-page\\">
    <div class=\\"lead\\">
        <h1>Client Application Permissions</h1>
        <p>Below is the list of applications you have given permission to and the resources they have access to.</p>
    </div>

        <div class=\\"row\\">
            <div class=\\"col-sm-8\\">
                <div class=\\"alert alert-info\\">
                    You have not given access to any applications
                </div>
            </div>
        </div>
</div>
    </div>

    <script src=\\"/lib/jquery/dist/jquery.slim.min.js\\"></script>
    <script src=\\"/lib/bootstrap/dist/js/bootstrap.bundle.min.js\\"></script>

    


</body></html>"
`;

exports[`Implicit Flow - user_with_all_claim_types - Introspection Endpoint 1`] = `
{
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_all_claim_types",
  "idp": "local",
  "some-app-user-custom-claim": "Oliver's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Oliver's Scope Custom User Claim",
  "amr": "pwd",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Implicit Flow - user_with_all_claim_types - UserInfo Endpoint 1`] = `
Object {
  "email": "oliver.hunter@gmail.com",
  "name": "Oliver Hunter",
  "some-custom-identity-user-claim": "Oliver's Custom User Claim",
  "sub": "user_with_all_claim_types",
}
`;

exports[`Implicit Flow - user_with_custom_api_resource_claims - Authorization Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_custom_api_resource_claims",
  "idp": "local",
  "some-app-user-custom-claim": "Sam's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Sam's Scope Custom User Claim",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ]
}
`;

exports[`Implicit Flow - user_with_custom_api_resource_claims - Grants 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, shrink-to-fit=no\\">

    <title>IdentityServer4</title>
    
    <link rel=\\"icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    <link rel=\\"shortcut icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    
    <link rel=\\"stylesheet\\" href=\\"/lib/bootstrap/dist/css/bootstrap.min.css\\">
    <link rel=\\"stylesheet\\" href=\\"/css/site.css\\">
</head>
<body>
    

<div class=\\"nav-page\\">
    <nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\">

        <a href=\\"/\\" class=\\"navbar-brand\\">
            <img src=\\"/icon.png\\" class=\\"icon-banner\\">
            IdentityServer4
        </a>

            <ul class=\\"navbar-nav mr-auto\\">
                <li class=\\"nav-item dropdown\\">
                    <a href=\\"#\\" class=\\"nav-link dropdown-toggle\\" data-toggle=\\"dropdown\\">user_with_custom_api_resource_claims <b class=\\"caret\\"></b></a>
                    
                    <div class=\\"dropdown-menu\\">
                        <a class=\\"dropdown-item\\" href=\\"/Account/Logout\\">Logout</a>
                    </div>
              </li>
            </ul>
    
    </nav>
</div>

   
    <div class=\\"container body-container\\">
        
<div class=\\"grants-page\\">
    <div class=\\"lead\\">
        <h1>Client Application Permissions</h1>
        <p>Below is the list of applications you have given permission to and the resources they have access to.</p>
    </div>

        <div class=\\"row\\">
            <div class=\\"col-sm-8\\">
                <div class=\\"alert alert-info\\">
                    You have not given access to any applications
                </div>
            </div>
        </div>
</div>
    </div>

    <script src=\\"/lib/jquery/dist/jquery.slim.min.js\\"></script>
    <script src=\\"/lib/bootstrap/dist/js/bootstrap.bundle.min.js\\"></script>

    


</body></html>"
`;

exports[`Implicit Flow - user_with_custom_api_resource_claims - Introspection Endpoint 1`] = `
{
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_custom_api_resource_claims",
  "idp": "local",
  "some-app-user-custom-claim": "Sam's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Sam's Scope Custom User Claim",
  "amr": "pwd",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Implicit Flow - user_with_custom_api_resource_claims - UserInfo Endpoint 1`] = `
Object {
  "email": "sam.tailor@gmail.com",
  "name": "Sam Tailor",
  "sub": "user_with_custom_api_resource_claims",
}
`;

exports[`Implicit Flow - user_with_custom_identity_claims - Authorization Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_custom_identity_claims",
  "idp": "local",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ]
}
`;

exports[`Implicit Flow - user_with_custom_identity_claims - Grants 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, shrink-to-fit=no\\">

    <title>IdentityServer4</title>
    
    <link rel=\\"icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    <link rel=\\"shortcut icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    
    <link rel=\\"stylesheet\\" href=\\"/lib/bootstrap/dist/css/bootstrap.min.css\\">
    <link rel=\\"stylesheet\\" href=\\"/css/site.css\\">
</head>
<body>
    

<div class=\\"nav-page\\">
    <nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\">

        <a href=\\"/\\" class=\\"navbar-brand\\">
            <img src=\\"/icon.png\\" class=\\"icon-banner\\">
            IdentityServer4
        </a>

            <ul class=\\"navbar-nav mr-auto\\">
                <li class=\\"nav-item dropdown\\">
                    <a href=\\"#\\" class=\\"nav-link dropdown-toggle\\" data-toggle=\\"dropdown\\">user_with_custom_identity_claims <b class=\\"caret\\"></b></a>
                    
                    <div class=\\"dropdown-menu\\">
                        <a class=\\"dropdown-item\\" href=\\"/Account/Logout\\">Logout</a>
                    </div>
              </li>
            </ul>
    
    </nav>
</div>

   
    <div class=\\"container body-container\\">
        
<div class=\\"grants-page\\">
    <div class=\\"lead\\">
        <h1>Client Application Permissions</h1>
        <p>Below is the list of applications you have given permission to and the resources they have access to.</p>
    </div>

        <div class=\\"row\\">
            <div class=\\"col-sm-8\\">
                <div class=\\"alert alert-info\\">
                    You have not given access to any applications
                </div>
            </div>
        </div>
</div>
    </div>

    <script src=\\"/lib/jquery/dist/jquery.slim.min.js\\"></script>
    <script src=\\"/lib/bootstrap/dist/js/bootstrap.bundle.min.js\\"></script>

    


</body></html>"
`;

exports[`Implicit Flow - user_with_custom_identity_claims - Introspection Endpoint 1`] = `
{
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_custom_identity_claims",
  "idp": "local",
  "amr": "pwd",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Implicit Flow - user_with_custom_identity_claims - UserInfo Endpoint 1`] = `
Object {
  "email": "jack.sparrow@gmail.com",
  "name": "Jack Sparrow",
  "some-custom-identity-user-claim": "Jack's Custom User Claim",
  "sub": "user_with_custom_identity_claims",
}
`;

exports[`Implicit Flow - user_with_standard_claims - Authorization Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_standard_claims",
  "idp": "local",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ]
}
`;

exports[`Implicit Flow - user_with_standard_claims - Grants 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
    <meta charset=\\"utf-8\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, shrink-to-fit=no\\">

    <title>IdentityServer4</title>
    
    <link rel=\\"icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    <link rel=\\"shortcut icon\\" type=\\"image/x-icon\\" href=\\"/favicon.ico\\">
    
    <link rel=\\"stylesheet\\" href=\\"/lib/bootstrap/dist/css/bootstrap.min.css\\">
    <link rel=\\"stylesheet\\" href=\\"/css/site.css\\">
</head>
<body>
    

<div class=\\"nav-page\\">
    <nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\">

        <a href=\\"/\\" class=\\"navbar-brand\\">
            <img src=\\"/icon.png\\" class=\\"icon-banner\\">
            IdentityServer4
        </a>

            <ul class=\\"navbar-nav mr-auto\\">
                <li class=\\"nav-item dropdown\\">
                    <a href=\\"#\\" class=\\"nav-link dropdown-toggle\\" data-toggle=\\"dropdown\\">user_with_standard_claims <b class=\\"caret\\"></b></a>
                    
                    <div class=\\"dropdown-menu\\">
                        <a class=\\"dropdown-item\\" href=\\"/Account/Logout\\">Logout</a>
                    </div>
              </li>
            </ul>
    
    </nav>
</div>

   
    <div class=\\"container body-container\\">
        
<div class=\\"grants-page\\">
    <div class=\\"lead\\">
        <h1>Client Application Permissions</h1>
        <p>Below is the list of applications you have given permission to and the resources they have access to.</p>
    </div>

        <div class=\\"row\\">
            <div class=\\"col-sm-8\\">
                <div class=\\"alert alert-info\\">
                    You have not given access to any applications
                </div>
            </div>
        </div>
</div>
    </div>

    <script src=\\"/lib/jquery/dist/jquery.slim.min.js\\"></script>
    <script src=\\"/lib/bootstrap/dist/js/bootstrap.bundle.min.js\\"></script>

    


</body></html>"
`;

exports[`Implicit Flow - user_with_standard_claims - Introspection Endpoint 1`] = `
{
  "iss": "http://localhost:8080",
  "aud": "some-app",
  "client_id": "implicit-flow-client-id",
  "sub": "user_with_standard_claims",
  "idp": "local",
  "amr": "pwd",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Implicit Flow - user_with_standard_claims - UserInfo Endpoint 1`] = `
Object {
  "email": "john.smith@gmail.com",
  "email_verified": "true",
  "name": "John Smith",
  "sub": "user_with_standard_claims",
}
`;
